enhancements/authentication/unauthenticated-subjects.md

```diff
@@ -0,0 +1,123 @@
+
+## Open Questions [optional]
+
+## Summary
+
+In earlier releases of OpenShift, the authorization module would allow anonymous users to access discovery endpoints. This enhancement proposes a way to remove unauthenticated subjects from cluster role bindings after an upgrade from this earlier release. It is possible that a customer is relying on anonymous user behavior in an implementation. This enhancement also proposes a way to opt-out of this reconciliation mechanism. Thus helping cases where changes are not desirable to customer implementations. Later releases of OpenShift removed this access for anonymous users. So, this enhancement would not affect the customers who install a fresh cluster using these later releases.
+
+## Motivation
+
+  * In earlier releases of OpenShift, an anonymous user could access “/” endpoint. Later releases revoked this access by removing unauthenticated subjects from cluster role bindings. Because of the way default policy resources are reconciled, after an upgrade from the earlier release, an anonymous user could still access "/" endpoint.
+  * Exposing unauthenticated endpoints to anonymous users can lead to information leakage. This is undesirable and needs prevention if possible. A user can see the list of available groups and endpoints on an API server using "/" endpoint. As this list could contain sensitive information, it is important to NOT allow anonymous users to access API Server discovery endpoints.
+
+### Goals
+
+> 1. In clusters upgraded from OCP 4.1, remove unauthenticated subjects from cluster role bindings with annotation "rbac.authorization.k8s.io/autoupdate=true".
+> 2. Users can opt-out of reconciliation if they set the annotation to "false".
+
+### Non-Goals
+
+## Proposal
+
+* Background
+  * A user who does not provide enough details for authentication with the kube-apiserver is considered anonymous and will be part of "system:unauthenticated" group.
+  * Both kube-apiserver and openshift-apiserver create a set of cluster roles and cluster role bindings during bootstrap.
+    * This [link](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#default-roles-and-role-bindings) has the list of default cluster roles and cluster role bindings in Kubernetes.
+    * This [link](https://docs.openshift.com/container-platform/4.4/authentication/using-rbac.html#default-roles_using-rbac) has the list of default cluster roles in OpenShift
+  * "rbac.authorization.k8s.io/autoupdate=true" annotation is also added to default policy resources.
+  * On start-up both openshift-apiserver and kube-apiserver reconcile default policy resources. They do the following operations based on the annotation
+    * if the annotation is not modified.
+        * add the missing permissions to cluster roles or add the missing subjects to cluster role bindings.
+    * if the annotation is deleted
+        * add the annotation back if it is missing
```

Comment:
if it is missing - isn't that implied by 'is deleted'?

Reply:
Yes, updated.